src/include/storage/index/bwtree_index.h

```diff
@@ -31,61 +28,182 @@ class BwTreeIndex final : public Index {
+                    const TupleSlot location) final {
     TERRIER_ASSERT(GetConstraintType() == ConstraintType::UNIQUE,
                    "This Insert is designed for indexes with uniqueness constraints.");
     KeyType index_key;
     index_key.SetFromProjectedRow(tuple, metadata_);
     bool predicate_satisfied = false;
 
-    // predicate is set to nullptr if the predicate returns true for some value
-    const bool ret = bwtree_->ConditionalInsert(index_key, location, predicate, &predicate_satisfied);
+    // The predicate checks if any matching keys have write-write conflicts or are still visible to the calling txn.
+    auto predicate = [&](const TupleSlot slot) -> bool {
+      const auto *const data_table = slot.GetBlock()->data_table_;
+      return data_table->HasConflict(*txn, slot) || data_table->IsVisible(*txn, slot);
+    };
+
+    const bool result = bwtree_->ConditionalInsert(index_key, location, predicate, &predicate_satisfied);
+
+    TERRIER_ASSERT(predicate_satisfied != result, "If predicate is not satisfied then insertion should succeed.");
 
-    // if predicate is not satisfied then we know insertion succeeds
-    if (!predicate_satisfied) {
-      TERRIER_ASSERT(ret, "Insertion should always succeed. (Ziqi)");
-    } else {
-      TERRIER_ASSERT(!ret, "Insertion should always fail. (Ziqi)");
+    if (result) {
+      // Register an abort action with the txn context in case of rollback
+      txn->RegisterAbortAction([=]() {
+        const bool UNUSED_ATTRIBUTE result = bwtree_->Delete(index_key, location);
+        TERRIER_ASSERT(result, "Delete on the index failed.");
+      });
     }
 
-    return ret;
+    return result;
+  }
+
+  void Delete(transaction::TransactionContext *const txn, const ProjectedRow &tuple, const TupleSlot location) final {
+    KeyType index_key;
+    index_key.SetFromProjectedRow(tuple, metadata_);
+
+    TERRIER_ASSERT(!(location.GetBlock()->data_table_->HasConflict(*txn, location)) &&
+                       !(location.GetBlock()->data_table_->IsVisible(*txn, location)),
+                   "Called index delete on a TupleSlot that has a conflict with this txn or is still visible.");
+
+    // Register a deferred action for the GC with txn manager. See base function comment.
+    auto *const txn_manager = txn->GetTransactionManager();
+    txn->RegisterCommitAction([=]() {
+      txn_manager->DeferAction([=]() {
+        const bool UNUSED_ATTRIBUTE result = bwtree_->Delete(index_key, location);
+        TERRIER_ASSERT(result, "Deferred delete on the index failed.");
+      });
+    });
   }
 
-  void ScanKey(const ProjectedRow &key, std::vector<TupleSlot> *value_list) final {
-    TERRIER_ASSERT(
-        value_list->empty(),
-        "Result set should begin empty. This can be changed in the future if index scan behavior requires it.");
+  void ScanKey(const transaction::TransactionContext &txn, const ProjectedRow &key,
```

Comment:
Since we are assuming the `value_list` is non-null, why not make it a reference?

Reply:
We use pointers for arguments when the object needs modifying within the function, and const reference otherwise.